src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites.
+        /// </summary>
+        public async Task<Solution> IntroduceParameterAsync(Document document, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, SyntaxNode containingMethod, bool allOccurrences, bool trampoline, bool overload,
+            CancellationToken cancellationToken)
+        {
+            var parameterName = await GetNewParameterNameAsync(document, expression, cancellationToken).ConfigureAwait(false);
+
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbol, cancellationToken).ConfigureAwait(false);
+
+            return await RewriteSolutionAsync(document,
+                expression, methodSymbol, containingMethod, allOccurrences, parameterName, methodCallSites,
+                trampoline, overload, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets the parameter name, if the expression's grandparent is a variable declarator then it just gets the
+        /// local declarations name. Otherwise, it generates a name based on the context of the expression.
+        /// </summary>
+        private static async Task<string> GetNewParameterNameAsync(Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName, out _))
+            {
+                return varDeclName;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(semanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Determines if the expression's grandparent is a variable declarator and if so,
+        /// returns the name
+        /// </summary>
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName, [NotNullWhen(true)] out SyntaxNode? localDeclaration)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
+            if (localDeclaration is null)
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+            {
+                varDeclName = null;
+                localDeclaration = null;
+                return false;
+            }
```

Comment:
sounds like a good case to add to the "could do in the future" list.

Reply:
added a comment indicating that